src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -339,117 +347,120 @@ infer' (Literal ss (ObjectLiteral ps)) = do
 
   fields <- forM ps inferProperty
   let ty = srcTypeApp tyRecord $ rowFromList (map toRowListItem fields, srcREmpty)
-  return $ TypedValue True (Literal ss (ObjectLiteral (map (fmap (uncurry (TypedValue True))) fields))) ty
+  return $ Inferred True (Literal ss (ObjectLiteral (map (fmap (uncurry (TypedValue True))) fields))) ty
 infer' (ObjectUpdate o ps) = do
   ensureNoDuplicateProperties ps
   row <- freshType
-  newVals <- zipWith (\(name, _) t -> (name, t)) ps <$> traverse (infer . snd) ps
+  newValsInferred <- zipWith (\(name, _) t -> (name, t)) ps <$> traverse (infer . snd) ps
   let toRowListItem = uncurry srcRowListItem
-  let newTys = map (\(name, TypedValue _ _ ty) -> (Label name, ty)) newVals
+  let newTys = map (\(name, Inferred _ _ ty) -> (Label name, ty)) newValsInferred
   oldTys <- zip (map (Label . fst) ps) <$> replicateM (length ps) freshType
   let oldTy = srcTypeApp tyRecord $ rowFromList (toRowListItem <$> oldTys, row)
-  o' <- TypedValue True <$> check o oldTy <*> pure oldTy
-  return $ TypedValue True (ObjectUpdate o' newVals) $ srcTypeApp tyRecord $ rowFromList (toRowListItem <$> newTys, row)
+  o' <- TypedValue True . inferredToExpr <$> check o oldTy <*> pure oldTy
```

Comment:
Bit of a nit but I think
```haskell
TypedValue True <$> (inferredToExpr <$> check o oldTy) <*> pure oldTy
```
might be a bit clearer here (they are equivalent, right?)

Reply:
They are indeed, and I think I used both forms in different places :) if you prefer the above it's fine with me